libbeat/processors/annotate/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/elastic/beats/libbeat/processors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	timeout = time.Second * 5
+)
+
+var (
+	fatalError = errors.New("Unable to create kubernetes annotator")
+)
+
+type kubernetesAnnotator struct {
+	podWatcher *podWatcher
+	logsPath   string
+}
+
+type kubeAnnotatorConfig struct {
+	InCluster  bool          `config:"in_cluster"`
+	ApiServer  string        `config:"api_server"`
+	KubeConfig string        `config:"kube_config"`
+	Host       string        `config:"host"`
+	Namespace  string        `config:"namespace"`
+	LogsPath   string        `config:"logs_path"`
+	SyncPeriod time.Duration `config:"sync_period"`
+}
+
+func init() {
+	processors.RegisterPlugin("annotate.kubernetes",
+		newKubernetesAnnotator)
+}
+
+func newKubernetesAnnotator(cfg common.Config) (processors.Processor, error) {
+
+	config := kubeAnnotatorConfig{
+		InCluster:  true,
+		SyncPeriod: 1 * time.Second,
+		Namespace:  "kube-system",
+		LogsPath:   "/var/lib/docker/containers/",
+	}
+
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("fail to unpack the annotate.kubernetes configuration: %s", err)
+	}
+
+	err = validate(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var c *rest.Config
+	if config.InCluster == true {
+		c, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, fmt.Errorf("Unable to get in cluster configuration")
+		}
+	} else {
+		c, err = clientcmd.BuildConfigFromFlags(config.ApiServer, config.KubeConfig)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to get cluster configuration")
+		}
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(c)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create kubeclient due to error: %v", err)
+	}
+
+	if config.Host == "" {
+		podName := os.Getenv("HOSTNAME")
+		logp.Info("Using pod name %s and namespace %s", podName, config.Namespace)
+		if podName == "localhost" {
+			config.Host = "localhost"
+		} else {
+			pod, error := kubeClient.CoreV1().Pods(config.Namespace).Get(podName, meta_v1.GetOptions{})
+			if error != nil {
+				logp.Err("Querying for pod failed with error: ", error.Error())
+				logp.Info("Unable to find pod, setting host to localhost")
+				config.Host = "localhost"
+			} else {
+				config.Host = pod.Spec.NodeName
+			}
+
+		}
+	}
+
+	logp.Info("Using host ", config.Host)
+	logp.Info("Initializing watcher")
+
+	if kubeClient != nil {
+		watcher := NewPodWatcher(kubeClient, config.SyncPeriod, config.Host)
+		if watcher.Run() == true {
+			return kubernetesAnnotator{podWatcher: watcher, logsPath: config.LogsPath}, nil
+		} else {
+			return nil, fatalError
+		}
+
+	}
+
+	return nil, fatalError
+}
+
+func (k kubernetesAnnotator) Run(event common.MapStr) (common.MapStr, error) {
+
+	metadata := common.MapStr{}
+	if value, ok := event["source"]; ok {
```

Comment:
my use case is that I have a source file that has a container id and i need to extract it out. This is not supported by format strings today.